File: ChatDemo-UI3.0/ChatDemo-UI3.0/Code/Controllers/Chats/Views/ChatCell/EMChatLocationBubbleView.m
/************************************************************
 *  * Hyphenate
 * __________________
 * Copyright (C) 2016 Hyphenate Inc. All rights reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of Hyphenate Inc.
 */

#import "EMChatLocationBubbleView.h"

#import "EMMessageModel.h"

#define LABEL_FONT_SIZE 13.f
#define LOCATION_IMAGEVIEW_SIZE 95

@interface EMChatLocationBubbleView ()

@property (nonatomic, strong) UILabel *addressLabel;

@end

@implementation EMChatLocationBubbleView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.layer.masksToBounds = YES;
        self.backImageView.image = [UIImage imageNamed:@"Location"];
        _addressLabel = [[UILabel alloc] init];
        _addressLabel.font = [UIFont systemFontOfSize:LABEL_FONT_SIZE];
        _addressLabel.textColor = AlmostBlackColor;
        _addressLabel.numberOfLines = 0;
        _addressLabel.backgroundColor = [UIColor clearColor];
        [self.backImageView addSubview:_addressLabel];
    }
    return self;
}

-(CGSize)sizeThatFits:(CGSize)size
{
    CGSize textBlockMinSize = {95, 25};
    EMLocationMessageBody *body = (EMLocationMessageBody*)self.model.message.body;
    CGSize addressSize = [body.address boundingRectWithSize:textBlockMinSize
                                                    options:NSStringDrawingUsesLineFragmentOrigin
                                                 attributes:@{NSFontAttributeName:_addressLabel.font}
                                                    context:nil].size;
    CGFloat width = addressSize.width < LOCATION_IMAGEVIEW_SIZE ? LOCATION_IMAGEVIEW_SIZE : addressSize.width;
    return CGSizeMake(width, LOCATION_IMAGEVIEW_SIZE);
}

-(void)layoutSubviews
{
    [super layoutSubviews];
    _addressLabel.frame = CGRectMake(5, self.backImageView.height - 30, self.backImageView.width - 10, 25);
}

- (void)setModel:(EMMessageModel *)model
{
    [super setModel:model];
    EMLocationMessageBody *body = (EMLocationMessageBody*)model.message.body;
    _addressLabel.text = body.address;
}

+ (CGFloat)heightForBubbleWithMessageModel:(EMMessageModel *)model
{
    return 100.f;
}

@end

File: swift-demo/Hyphenate-Demo-Swift/Code/Controllers/Chats/Views/ChatCell/EMChatLocationBubbleView.swift
//
//  EMChatLocationBubbleView.swift
//  Hyphenate-Demo-Swift
//
//  Created by dujiepeng on 2017/6/26.
//  Copyright © 2017 dujiepeng. All rights reserved.
//

import UIKit
import Hyphenate

let LOCATION_IMAGEVIEW_SIZE: CGFloat = 95

class EMChatLocationBubbleView: EMChatBaseBubbleView {
    var addressLabel: UILabel?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        layer.masksToBounds = true
        _backImageView?.image = UIImage(named:"Location")
        addressLabel = UILabel()
        addressLabel?.textColor = AlmostBlackColor
        addressLabel?.font = UIFont.systemFont(ofSize: LABEL_FONT_SIZE)
        addressLabel?.numberOfLines = 0
        addressLabel?.backgroundColor = UIColor.clear
        _backImageView?.addSubview(addressLabel!)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        let textBlockMinSize: CGSize = CGSize(width: 95, height: 25)
        let body = _model?.message?.body as! EMLocationMessageBody
        let addressSize = (body.address!.boundingRect(with: textBlockMinSize, options: NSStringDrawingOptions.usesLineFragmentOrigin, attributes: [NSFontAttributeName:addressLabel!.font], context: nil) as CGRect).size

        let width = addressSize.width < LOCATION_IMAGEVIEW_SIZE ? LOCATION_IMAGEVIEW_SIZE : addressSize.width
        
        return CGSize(width: width, height: LOCATION_IMAGEVIEW_SIZE)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        addressLabel?.frame = CGRect(x: 5, y: (_backImageView?.height())! - 30, width: (_backImageView?.width())! - 30, height: 25)
    }
    
    override func set(model: EMMessageModel) {
        super.set(model: model)
        let body = _model?.message?.body as! EMLocationMessageBody
        addressLabel?.text = body.address
    }
    
    class func heightForBubble(withMessageModel model: EMMessageModel) -> CGFloat {
        return 100
    }
}
